File: Resilience.hs
{-# LANGUAGE OverloadedStrings#-}

import System.Environment(getArgs)
import System.IO(openFile,
                 hClose,
                 IOMode(..),
                 hPutStrLn)

--import qualified Data.ByteString as B
--import qualified Data.Text.Encoding as E

import qualified Data.Text as T
import qualified Data.Text.IO as TI

import Data.Array.Unboxed(Array,
                          (!))
import qualified Data.Vector as V
import Data.Ratio(Ratio,
                  (%))

--for stateful programs
import Data.Array.ST(runSTArray,
                     newArray,
                     writeArray)
import Control.Monad(forM,
                     forM_)
--import Control.Monad.ST

--Sample values
a :: Integer
a = 261

b :: Integer
b = 1537

q :: Ratio Integer
q = a%b

toQ :: (T.Text,T.Text) -> Ratio Integer
toQ (a,b) = (read (T.unpack a))%(read (T.unpack b))

textToTuples :: T.Text -> (T.Text,T.Text)
textToTuples text = (res !! 0, res !! 1)
  where
    res = T.words text

listTTolistT :: [T.Text] -> [(T.Text,T.Text)]
listTTolistT vals = map textToTuples vals

toListQ :: [(T.Text,T.Text)] -> [Ratio Integer]
toListQ vals = map toQ vals

--list of primes
sieve :: [Integer] -> [Integer]
sieve [] = []
sieve (nextPrime:rest) = nextPrime : sieve noFactors
  where noFactors = filter (not . (== 0) . (`mod` nextPrime)) rest

primes :: [Integer]
primes = sieve [2 .. ]

--totient
totient :: [Integer] -> Integer
totient prseq = foldl (*) 1 newseq
  where
    newseq = map (+ (-1)) prseq

resList :: Ratio Integer -> [Integer] -> [Integer] -> Integer -> Integer -> Integer -> Integer -> ([Integer],Integer,Integer)
resList val lval lprimes mult oldmult tot oldtot = if tot%(mult - 1) >= val
  then resList val newList newPrimes newMult mult a tot
  else (lval,oldmult,oldtot)
  where
    newList = (head lprimes):lval
    newPrimes = tail lprimes
    newMult = mult*(head lprimes)
    a = tot*((head lprimes) - 1)

listToSearch :: (a,b,c) -> a
listToSearch (x,_,_) = x

oldMul :: (a,b,c) -> b
oldMul (_,y,_) = y

oldT :: (a,b,c) -> c
oldT (_,_,z) = z

nR :: [Ratio Integer] -> Integer -> Integer -> Integer -> [Ratio Integer]
nR lvals 1 _ _ = lvals
nR lvals factor oldmult oldtot = nR (x:lvals) (factor - 1) oldmult oldtot
  where
    x = (oldtot*factor)%(factor*oldmult - 1)

listToArray :: Integer -> [Ratio Integer] -> Array Integer (Ratio Integer)
listToArray size vals = runSTArray $ do
    myArray <- newArray (2,size) 0
    forM_ (zip vals [2 .. ]) $ \(val,i) -> writeArray myArray i val
    return myArray

binarySearch :: (Ratio Integer) -> Integer -> Integer -> Array Integer (Ratio Integer) -> Integer
binarySearch qVal fstM lastM myArray
  | fstM == lastM = fstM
  | value < qVal = binarySearch qVal fstM index myArray
  | otherwise = binarySearch qVal (index + 1) lastM myArray
  where
     index = (lastM + fstM) `quot` 2
     value = myArray ! index

fullTransform :: (Ratio Integer) -> Integer
fullTransform qVal = finalIndex * oldmult
  where
    tripleM = resList qVal [2] (tail primes) 2 2 1 1
    oldmult = oldMul tripleM
    finalIndex = binarySearch qVal 2 size myArray
    size = head $ listToSearch tripleM
    myArray = listToArray size vals
    vals = nR [] size oldmult oldtot
    oldtot = oldT tripleM

main :: IO ()
main = do
  --getting contents from the file
  args <- getArgs
  let fileName = head args
  file <- openFile fileName ReadMode
  input <- TI.hGetContents file
  hClose file
  --prepare contents for functions
  let readableInput = T.lines input
  let numT = head readableInput
  lineResults <- forM (take (read (T.unpack numT)) (tail readableInput)) return
  let listOfQs = toListQ $ listTTolistT lineResults
  --finding the answers
  let listOfAs = map fullTransform listOfQs
  let listOfAsStr = map show listOfAs
  --printing to a file
  fileW <- openFile "Results.txt" WriteMode
  mapM_ (hPutStrLn fileW) listOfAsStr
  hClose fileW
  print "Done"

File: Resilience.py
import math
import fractions
import time
import array as ar

n = 258

a = time.time_ns()

primes = ar.array('I',[i for i in range(2,n + 1)])
primes_bool = ar.array('I',[True for i in range(2,n + 1)])

for i in range(0,int(math.sqrt(n + 1))):
    if primes_bool[i] == True:
        for j in range((i + 2)*(i + 2) - 2, (n - 1), (i + 2)):
            primes_bool[j] = False

k = 0

for i in range(0, n - 1):
    if not(primes_bool[i]):
        primes.pop(k)
        k = k - 1
    k = k + 1

b = time.time_ns()

#Res List
def resilience_list(val):
    mult = 2
    old_mult = 2
    tot = 1
    old_tot = 1
    i = 1

    while fractions.Fraction(tot,(mult - 1)) >= val:
        old_mult = mult
        old_tot = tot
        mult = mult*primes[i]
        tot = tot*(primes[i] - 1)
        i = i + 1

    return (i,old_mult,old_tot)

def binary_search(list_to_search,q_value,old_mult,old_tot):
    min = 0
    max = len(list_to_search) - 1
    while not(min == max):
        index = (min + max) // 2
        value = fractions.Fraction(old_tot*(list_to_search[index]),(old_mult*(list_to_search[index]) - 1))
        if value < q_value:
            max = index
        else:
            min = index + 1
    return max

def integer_answer(tonv,val):
    lts = [(i + 1) for i in range(1,primes[tonv[0] - 1])]
    m = binary_search(lts,val,tonv[1],tonv[2])
    int_answer = tonv[1]*lts[m]
    return int_answer

c = time.time_ns()

list_of_Qs = []
num_of_lines_to_read = int(input())
for l in range(1,num_of_lines_to_read + 1):
    #lines are in the form 'int int'
    num = input()
    #separate them to ['int','int']
    nums = num.split()
    #create fractions
    list_of_Qs.append(fractions.Fraction(int(nums[0]),int(nums[1])))

list_of_answers = []
for l in range(1,num_of_lines_to_read + 1):
    val = list_of_Qs[l - 1]
    ans = integer_answer(resilience_list(val),val)
    list_of_answers.append(ans)

for l in range(1,num_of_lines_to_read + 1):
    print(list_of_answers[l - 1])

d = time.time_ns()

print('b - a =',(b - a)/1000000000)
print("{} = {}".format("d - c",(d - c)/1000000000))
print("{} = {}".format("d - a",(d - a)/1000000000))
